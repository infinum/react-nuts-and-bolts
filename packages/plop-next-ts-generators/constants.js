const path = require('path');
const { getFoundations } = require('./utils');

const dir = path.resolve(__dirname);

const CORE_DOMAIN = 'core';
const SHARED_DOMAIN = 'shared';
const FEATURE_DOMAIN = 'features';
const rootDomains = [CORE_DOMAIN, SHARED_DOMAIN, FEATURE_DOMAIN];

const SINGLE_PART_COMPONENT = 'single-part';
const MULTI_PART_COMPONENT = 'multi-part';
const componentTypes = [SINGLE_PART_COMPONENT, MULTI_PART_COMPONENT];

const INIT_COMMAND = 'init';
const ADD_FOUNDATION_COMMAND = 'add-foundation';
const ADD_COMPONENT_COMMAND = 'add-component';
const foundationCommands = [INIT_COMMAND, ADD_FOUNDATION_COMMAND, ADD_COMPONENT_COMMAND];

const RADIUS_FOUNDATION = 'radius';
const SPACING_FOUNDATION = 'spacing';
const TYPOGRAPHY_FOUNDATION = 'typography';
const Z_INDEX_FOUNDATION = 'z-index';

const foundations = getFoundations({ dir });

/**
 * @constant
 * @type {import('./index').Config}
 */
const defaultConfig = {
	baseUrl: 'src',
	componentsUrl: 'components',
};

module.exports = {
	dir,
	defaultConfig,

	CORE_DOMAIN,
	SHARED_DOMAIN,
	FEATURE_DOMAIN,
	rootDomains,

	SINGLE_PART_COMPONENT,
	MULTI_PART_COMPONENT,
	componentTypes,

	INIT_COMMAND,
	ADD_FOUNDATION_COMMAND,
	ADD_COMPONENT_COMMAND,
	foundationCommands,
	foundations,
	RADIUS_FOUNDATION,
	SPACING_FOUNDATION,
	TYPOGRAPHY_FOUNDATION,
	Z_INDEX_FOUNDATION,
};
