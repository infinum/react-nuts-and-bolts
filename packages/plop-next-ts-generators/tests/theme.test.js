import { resolve } from 'node:path';
import * as fs from 'node:fs';
import { render, waitFor } from 'cli-testing-library';

import { getFileHelper } from './file-helper.js';
import { ADD_FOUNDATION_COMMAND, INIT_COMMAND } from '../constants.js';

describe('theme generators', () => {
	const { getFilePath } = getFileHelper();

	/** @type {import('cli-testing-library').RenderResult} */
	let cli;

	beforeEach(async () => {
		cli = await render('plop', [
			`--plopfile=${resolve(__dirname, './plopfile.js')}`,
			`--dest=${resolve(__dirname, './output')}`,
			'theme',
		]);
	});

	it('should init theme', async () => {
		const themeFilePath = await getFilePath(`./output/src/styles/theme/index.ts`);

		const { findByText, userEvent } = cli;

		expect(await findByText('Select a command:')).toBeInTheConsole();

		expect(await findByText(`❯ ${INIT_COMMAND}`)).toBeInTheConsole();

		userEvent.keyboard('[Enter]');

		await waitFor(async () => fs.promises.stat(themeFilePath));

		const theme = fs.readFileSync(themeFilePath, 'utf8');

		expect(theme).toMatchSnapshot();
	});

	it('should add colors foundation theme', async () => {
		const themeFilePath = await getFilePath(`./output/src/styles/theme/index.ts`);
		const colorsFoundationFilePath = await getFilePath(`./output/src/styles/theme/foundations/colors.ts`);

		const { findByText, userEvent, getStdallStr } = cli;

		// Init theme file first
		const themeInitializer = await render('plop', [
			`--plopfile=${resolve(__dirname, './plopfile.js')}`,
			`--dest=${resolve(__dirname, './output')}`,
			'theme',
		]);

		expect(await themeInitializer.findByText('Select a command:')).toBeInTheConsole();

		expect(await themeInitializer.findByText(`❯ ${INIT_COMMAND}`)).toBeInTheConsole();

		themeInitializer.userEvent.keyboard('[Enter]');

		await waitFor(async () => fs.promises.stat(themeFilePath));

		// Add colors foundation

		expect(await findByText('Select a command:')).toBeInTheConsole();

		userEvent.keyboard('[ArrowDown]');

		expect(await findByText(`❯ ${ADD_FOUNDATION_COMMAND}`)).toBeInTheConsole();

		userEvent.keyboard('[Enter]');

		expect(await findByText('Select foundation:')).toBeInTheConsole();

		userEvent.keyboard('[ArrowDown]');
		userEvent.keyboard('[ArrowDown]');
		userEvent.keyboard('[ArrowDown]');

		expect(await findByText(`❯ colors`)).toBeInTheConsole();

		await userEvent.keyboard('[Enter]');

		await waitFor(async () => fs.promises.stat(colorsFoundationFilePath));

		console.log(getStdallStr());

		const colorFoundation = fs.readFileSync(colorsFoundationFilePath, 'utf8');
		const theme = fs.readFileSync(themeFilePath, 'utf8');

		expect(colorFoundation).toMatchSnapshot();
		expect(theme).toMatchSnapshot();
	});
});
