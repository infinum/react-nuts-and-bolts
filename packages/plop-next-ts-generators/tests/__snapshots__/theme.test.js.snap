// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`theme generators should add colors foundation theme 1`] = `
"export const colors = {
  // Add new color gradient token, for example:
	// primary: {
	//	50: '#f6faff',
	//	100: '#d7e9fd',
	//	200: '#c0dcfc',
	//	300: '#a7cefb',
	//	400: '#8abdfa',
	//	500: '#68aaf8',
	//	600: '#3d92f6',
	//	700: '#0070f3',
	//	800: '#0064d8',
	//	900: '#0055b9',
	//},
};
"
`;

exports[`theme generators should add colors foundation theme 2`] = `
"import { extendTheme } from '@chakra-ui/react';

import { colors } from './foundations/colors';
// -- PLOP:IMPORT_FOUNDATION_THEME --

// -- PLOP:IMPORT_COMPONENT_THEME --

const overrides = {
	// -- PLOP:REGISTER_FOUNDATION_THEME --
	components: {
		// -- PLOP:REGISTER_COMPONENT_THEME --
	},
};

const theme = extendTheme(overrides);

export default theme;
"
`;

exports[`theme generators should init theme 1`] = `
"import { extendTheme } from '@chakra-ui/react';

// -- PLOP:IMPORT_FOUNDATION_THEME --

// -- PLOP:IMPORT_COMPONENT_THEME --

const overrides = {
	// -- PLOP:REGISTER_FOUNDATION_THEME --
	components: {
		// -- PLOP:REGISTER_COMPONENT_THEME --
	},
};

const theme = extendTheme(overrides);

export default theme;
"
`;
